New features

nlemoine opened the pull request:
#1 - Enable fake data deletion
#7 - Add fake post_thumbnail data

nlemoine committed:
#7 - Add fake post_thumbnail data
#1 - Enable fake data deletion

nlemoine commented:
Thanks for this awesome package. 

Let me know if you want to change something.

haroldangenent commented on src/Command.php:
```diff
@@ -23,4 +23,88 @@ public function fake(array $args = [])
+     * Delete attachments
+     * @return array Array of deleted attachment IDs
+     */
+    private static function deleteAttachments()
```
I'd rather not use static methods (not necessary anymore when the business logic is separated from the command).

haroldangenent commented on src/Command.php:
```diff
@@ -23,4 +23,88 @@ public function fake(array $args = [])
+
+        // Delete posts
+        foreach ($postIds as $postId) {
+            if( $type === 'post' ) {
```
This implies only `post` post types can be removed. However, Faker can be used to generate any post type.

nlemoine replied:
This actually doesn't match`post` post types. Just the method parameter to distinguish posts from attachments.

haroldangenent commented on src/Provider/PostThumbnail.php:
```diff
@@ -0,0 +1,41 @@
+
+final class PostThumbnail
+{
+    public function thumbnail($width = 640, $height = 480, $category = null)
```
Maybe increase default sizes, since this will be left empty on most cases (and WordPress will generate them).

haroldangenent commented on src/Provider/PostThumbnail.php:
```diff
@@ -0,0 +1,41 @@
+		// Assign metadata to attachment
+		wp_update_attachment_metadata( $attachmentId,  $attachmentData );
+
+		update_post_meta($attachmentId, '_fake_attachment', true);
```
Why not use the same meta key here (`_fake`)? :)

nlemoine replied:
So i can identify attachments from posts. Deletion is a bit different for attachments.

haroldangenent reviewed (changes requested):
@nlemoine Thanks for your work on this! I didn't get a change to test this, but I have some superficial feedback for now.
- Code does not seem to be PSR-4. You can run the gulpfile in the repo that includes PHP Code Sniffer. Use `npm install` for this.
- Both new features do not include tests.
- Most deletion code has been added to `Command.php`, but this file is not intended to contain business logic (it's used as a Middle Man between CLI and business logic).
- I'm not sure about your approach for generating thumbnails. This is currently a provider for posts, but maybe attachments could be an entity of their own, that can be referenced in any of the fields. This would make it much easier to set up custom fields with images or maybe even other types of files and would treat Attachment as just another post type. See [this example](https://github.com/nelmio/alice/#example) for a reference to another entity. I believe this approach would be more powerful.

nlemoine commented:
> Code does not seem to be PSR-4. You can run the gulpfile in the repo that includes PHP Code Sniffer. Use npm install for this.
> Both new features do not include tests.
> Most deletion code has been added to Command.php, but this file is not intended to contain business logic (it's used as a Middle Man between CLI and business logic).

Sorry about that, I made this a little bit too quick as I needed this feature for my own needs. 

> I believe this approach would be more powerful.

Working on it after my PR, I also thought to myself it could be much better that way.  But there's an issue I'm a bit struggling with.

You need the ID of the attachment in order to make a reference in a post such as `'@attachment*->ID'`
To get an ID, attachment need to be inserted in the database. 
But all entities are being processed before they are persisted in database by Alice so no way the get the attachment ID. 

I managed to get it working by splitting yaml files (and changing a bit the command file so it can accept multiple files). Putting the attachments file first, this is ok. 

But I don't really like it that way. So I came to another solution. Save a temporary post in the entity constructor, something like: 

``` php
function __contruct() {
  $this->ID = wp_insert_post(['post_title'=>'temp']);
}
```

So the entities have a real ID since object creation. 
Then I created a `Persister` class (see https://github.com/nelmio/alice/blob/2.2.2/src/Nelmio/Alice/PersisterInterface.php) that updates each post with fake generated values. 

Still working on it but I made some major changes to your repo. I might release another package for this.

haroldangenent commented:
> To get an ID, attachment need to be inserted in the database. 
> But all entities are being processed before they are persisted in database by Alice so no way the get the attachment ID.

Ah, of course. WordPress generates the post ID. I hope I'll find some time to look into this a bit deeper next week. Thank you for explaining your thought process.

haroldangenent commented:
I dove a little deeper into generating attachments. I realized this would neem some refactoring, to set up an environment where:
1. Multiple types of entities can be created and maintained
2. The command could take multiple files

I chose to go down the path that takes multiple files as an argument, because that's what [nelmio/alice's documentation](https://github.com/nelmio/alice/blob/2.x/doc/relations-handling.md#references) suggests.

I have created another PR for this in #10. Hope you don't mind. I'm keen to know what you think about this approach.

nlemoine commented:
Thanks for your feedback. No problem for the PR. These features sure take some refactoring. 

I'm currently working on it too but with the constraint of keeping everything in a single file. I think maintaining multiple files and loading them in a predefined order is not developer friendly enough. 

Will show you something as soon as it is stable enough.

haroldangenent commented:
I'll close this PR for now in favor of the features in the new release. I'm still curious about your solution, let me know if you find the time for this.

nlemoine commented:
Almost finished. I'll let you know as soon as I pushed something.

nlemoine commented:
https://github.com/nlemoine/wp-cli-fixtures

There it is. Still not submitted to wp cli package index. I'm going to have lunch. Will explain some choices when I come back.

nlemoine commented:
1. As explained earlier, having multiple files wasn't an option. I wanted to deal with everything in a single yaml file.
2. I added the ability to generate any type of WordPress content (posts, attachments, users, comments, terms).
3. I revised the way you deal with images. Actually, images in Faker are saved in a temp folder and then duplicated in in your WordPress uploads folder so images are existing twice on your drive. Also you can't pass options to `<image()>` (because first argument is the target directory) such as width, height or category.
4. I added the ability to delete custom fake content. 
5. I provided some custom providers to return an existing post, attachment, term or user ID: https://github.com/nlemoine/wp-cli-fixtures/blob/master/src/Provider/WordPress.php
You can pass some query arguments to these provider (WP_Query, WP_User_Query, etc.).

nlemoine commented:
Forgot some changes:
6. Ability to add fake content to existing entities.
7. Uses the latest version of Alice.

haroldangenent commented:
@nlemoine You seem to have put a lot of work and thought in this and the project seems to have a lot of cool added features!

While I don't agree with all choices you've made in the project, I understand why you've made most of them. Just by scanning the code real quickly, the greatest difference is probably that I didn't want to add an abstract layer in between `wp-cli` and `alice`, to keep the code simple and as maintainable as possible for future changes and features (without having to invest a lot of time to maintain it in a backwards compatible fashion).

Another benefit of this approach is that the inner workings of the package are easier to understand for users, which will make debugging and collaboration a bit easier.

By the way, I see you have removed all integration tests (and didn't create ones for the new feature). To create a stable and maintainable package, writing tests would be a great addition to your package.

**Multiple wp-cli packages for the same goal**
Lately, I have seen some discussion about multiple packages for the same goal in [this topic](https://github.com/wp-cli/package-index/pull/90) and especially [this comment](https://github.com/wp-cli/package-index/pull/90#issuecomment-258513610).

Your package seems to add some cool features for end users (which were still on the to-do list for our repo). While I (really) get that you've created another repository, it might be more beneficial for those users if there was one package that handles a task.

I would love for these features to be able to merge into this package (or at least be created for this package). While the codebase of your package seems to be heavily based upon this package, it does not seem like an easy merge.

**Workflow for collaborating**
A great and tested way to collaborate on such projects would be to create multiple concise Pull Requests (that handle a very specific problem or introduce a very specific feature), that contain integration tests for these features or changeset.

This would allow us to review this specific changeset and to improve the repository in a stable way. This workflow seems to work for a lot of projects and is how we currently work on this project.

Please don't feel forced to collaborate on 'my package' and feel free to make the decision that works best for you. I'm sure the community will benefit either way.

nlemoine commented:
Thanks for your detailed feedback! 

I also read the comment you mentionned and have some mixed feelings about it. I totally understand the wp cli maintainers arguments, It makes sense that they don't want to end up like WordPress plugins repo, with thousands of plugins providing the same feature or unmaintained repos. On the other hand, the diversity of implementations, designs is what makes some repository great and provides various choices to developers. I believe wp cli is a bit different from WordPress plugins. It mostly target WordPress developers, not end WordPress users and it's probably less profitable business to write wp cli packages than freemium plugins. 

I guess It would be great to have a single package but like you said, it won't be an easy merge. I already spent some time rewriting this package (already made the same options you took before your last changes: moving entities to a separate folder, wrote a persister class, as mentionned in https://github.com/trendwerk/faker/pull/8#issuecomment-255368597) and adding new features, I won't spend more to match your package design. Merging projects would first require to agree on some design options (alice version to use, single vs multiple yaml files, etc.). Developers often work in an opinionated way. I rewrote most parts of this project for this reason. Also, in the first place, I added all these features to quickly match my needs for a project. It's ok to disagree, I don't really mind if my package doesn't end up in the wp cli package index ;)

Let me know if you want to discuss it further.

haroldangenent commented:
Thanks for your explanation. This makes me understand where you're coming from a lot better.

> I rewrote most parts of this project for this reason. Also, in the first place, I added all these features to quickly match my needs for a project.

Considering those requirements, the best choice for you would indeed be to create a fork or your own package. Spending a lot of time for a complex merger and / or discussing all possible directions in which the package could go to seems like a waste of time for those needs.

> Developers often work in an opinionated way. [..] It's ok to disagree

You're definitely right. I'm also not sure how and in what direction the `wp-cli` package index will grow, but the ideas outlined by Daniel (i.e. becoming more like [The PHP League](https://thephpleague.com/)) seem like a good ambition.

In this case the best thing might be to keep the repositories separated, considering the large amount of time you and I will have to invest to merge them and discuss our differences and the goals you have outlined in your comment.
